### PR TITLE
[FLINK-33667] Implement restore tests for MatchRecognize node

### PR DESCRIPTION
**Pull request by jnh5y**
## What is the purpose of the change

Implement restore tests for MatchRecognize node

## Verifying this change

This change added tests and can be verified as follows:

* Added restore tests for MatchRecognize node which verifies the generated compiled plan with the saved compiled plan

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Review by dawidwys — COMMENTED** [review]
Looks good. Added suggestion to extend test coverage.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/MatchRecognizeTestPrograms.java [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    static final TableTestProgram COMPLEX_MATCH =
+            TableTestProgram.of("complex-match", "complex match recognize test")
+                    .setupTableSource(
+                            SourceTestStep.newBuilder("MyTable")
+                                    .addSchema(
+                                            "symbol string",
+                                            "tstamp bigint",
+                                            "price int",
+                                            "tax int",
+                                            "proctime as PROCTIME()")
```
[the remark this answers is not in the document]
Comment (jnh5y) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/MatchRecognizeTestPrograms.java:95]: Sure.  Does MR do anything different depending on rowtime vs proctime?

Also, just to make sure, what do you mean be secondary sorting?  Are you suggesting sorting "outside" the MR clause?

[the remark this answers is not in the document]
Comment (jnh5y) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/MatchRecognizeTestPrograms.java:95]: Nice; thanks for the suggestion.  I've added two tests in https://github.com/apache/flink/pull/23821/commits/74f3b69430d7f55fb385d1463e71dbbb8142a29f.  Lemme know if you can think of other cases to check out.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/MatchRecognizeTestPrograms.java [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+                                    + "             PATTERN (`A\"` \u006C C)\n"
+                                    + "             DEFINE\n"
+                                    + "                 `A\"` AS name = 'a',\n"
+                                    + "                 \u006C AS name = 'b',\n"
+                                    + "                 C AS name = 'c'\n"
+                                    + "     ) AS T")
+                    .build();
+
+    static final TableTestProgram COMPLEX_MATCH =
+            TableTestProgram.of("complex-match", "complex match recognize test")
```
Comment (dawidwys) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/MatchRecognizeTestPrograms.java:87]: Could you please adapt the id according to the javadoc:
https://github.com/apache/flink/blob/52d8d3583e5c989da84126a8805ab335408c46c2/flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/test/program/TableTestProgram.java#L117
?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/MatchRecognizeTestPrograms.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/MatchRecognizeTestPrograms.java:95]: Thank you!

**Review by dawidwys — APPROVED** [review]
LGTM, I'll merge the PR after https://github.com/apache/flink/pull/23869 which is necessary for the cleanup.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/MatchRecognizeTestPrograms.java (hunk 2 above)
Comment (jnh5y) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/MatchRecognizeTestPrograms.java]: Ah, yes; didn't see that.  I was reusing identifiers from the existing tests!

Updated.

**Comment by jnh5y** [comment]
> LGTM % the ids issue, I'll merge the PR after #23869 which is necessary for the cleanup.

FWIW, MR observes the changelog not final results.  I *think* it should be ok to merge before or after the Join PR.

**Pull request merged**